### ISSUE TITLE
Trouble with training with multiple inputs

### ISSUE THREAD

**Issue by opfromthestart** [coreylowman/dfdx#573]
I have a specific use case that does not appear to work with the current architecture, and I would like some help figuring out how to implement it.
I have a network that takes two inputs, an image and a vector. The network needs to apply convolutions to the image, then reshape the output from the convolution, concatenate it with the input vector, and pass that into linear layers. This does not work with the current model as I cannot pass an object like `(&part1, &part2)` into optimizers, and if they are handled separately I get `UnusedParams` errors when trying to update. If I use 2 separate `Gradient`s, the program compiles and runs but neither part learns.
Any help on this would be appreciated.

Code attached, look in `train` and `forward_raw` for specifics.
[code.zip](https://github.com/coreylowman/dfdx/files/10986684/code.zip)

**Comment by TimerErTim** [comment]
You should be able to achieve this by implementing a `TensorCollection` for your `NetCrit`. You can then pass the whole `TensorCollection` into the optimizer.  And you can allocate all gradients using that collection. I think that is the way to go.

I have not worked with that tho, so I am only theorizing here.

**Comment by opfromthestart** [comment]
This code still gives a `UnusedParams(UnusedTensors { ids: [UniqueId(6), UniqueId(7), UniqueId(8), UniqueId(9), UniqueId(10), UniqueId(11)] })` error because I cannot call `update` on both without one of them erroring.

**Comment by coreylowman** [comment]
Ahh yes I see! I'd suggest:

1. Combine self.layer_lin & self.layer_conv into a single field
```rust
pub(crate) struct NetCrit {
    net: (<NetConv as BuildOnDevice<Device, f32>>::Built, <NetLinCrit as BuildOnDevice<Device, f32>>::Built),
}
```

2. Use a single Sgd object for both
```rust
let mut sgd : Sgd<_,_,Device> = Sgd::new(&self.net, SgdConfig{
            lr: 0.1,
            momentum: Some(dfdx::optim::Momentum::Nesterov(0.9)),
            weight_decay: None,
        });
```

3. And then you'd only need one call to .update()
```rust
sgd.update(&mut self.net, &grads).unwrap();
```

**Comment by opfromthestart** [comment]
That worked to allow it to only use one optimizer, but the model still does not improve after training.
[code.zip](https://github.com/coreylowman/dfdx/files/10993025/code.zip)

**Comment by coreylowman** [comment]
There's a lot of things that could be going wrong if the model error is not decreasing as you expect. Things like:
1. too large of a learning rate
2. not enough data or ill-formed data
3. batch size not tuned
4. hyperparameters not tuned correctly
5. Probably a ton of other stuff

If you've implemented this exact flow in another framework and the performance doesn't match, then I can likely help here. Otherwise I don't have enough info at this point.

**Comment by opfromthestart** [comment]
I was stupid, I had a bug in previous code that made it look like it wasn't learning when it actually was. Thank you.
